# Calculos para desfasaje PWM PCA9685

In [17]:
import numpy as np

## Primer PWM  
El primer PWM incia con delay del 0%, con un ciclo ultil del 20% y tiempo de apagado de 80%:
Tiempo de retardo 0% = 0 counts = 0x0  

LED0_ON_H = 0x0;  
LED0_OFF_L = 0x0;  

Tiempo de encendido 80% -> 3276 counts, el dato a escribir en los registros LED_OFF_H y LED_OFF_L son los siguientes

In [29]:
resolution = 4096
value1 = resolution * .80
print(value1)
LEDOFF = round(value1)-1
print(LEDOFF)
print("Valor a escribir en el registro LED_OFF:", np.base_repr(LEDOFF,16))

3276.8
3276
Valor a escribir en el registro LED_OFF: CCC


LED_OFF_H = 0xC;  
LED_OFF_L = 0xCC;

## Segundo PWM  
El segundo PWM incia con un delay del 20%, un ciclo util del 80% y tiempo de apagado de 20%;  
El delay time es de 819 counts restando 1

In [19]:
value2 = resolution * .20
LED1ON = round(value2)-1
print(LED1ON)
np.base_repr(LED1ON,16)

818


'332'

Este valor sera escrito en los registros LED1_ON  
LED1_ON_H = 0x3;  
LED1_ON_L = 0x32;  
EL tiempo encendido sera entonces de 80% -> 3276.8 counts por lo que LED1_OFF = ((819 + 3276.8)-1)

In [20]:
value3 = resolution * .80
LED1OFF=((round(value2)+round(value3))-1)
print(LED1OFF)
np.base_repr(LED1OFF,16)

4095


'FFF'

El dato a escribir en los registros LED1_OFF_H y LED1_OFF_L son los siguientes  
LED1_OFF_H=0xF;  
LED1_OFF_L=0xFf;

## Tercer PWM  
Los valores del registro LED2_ON incian en el valor anterior con un delay time del 40%, un ciclo util del 20% y un tiempo de apagado del 80%;  
LED2ON=round(4096*.40)-1  

In [21]:
value4 = resolution * .40
LED2ON=round(value4)-1
print(LED2ON)
np.base_repr(LED2ON,16)

1637


'665'

LED2_ON_H = 0x6;  
LED2_ON_L = 0x65;  
EL tiempo encendido sera entonces de 80% -> 3276.8 counts con un apagado de 20% por lo que LED2_OFF = ((1637+3276)=val-4096)

In [22]:
value5 = resolution * .80
LED2OFF=((round(LED2ON)+round(value5))-4096)
print(LED2OFF)
np.base_repr(LED2OFF,16)

818


'332'

Los valores hexadecimales anteriores son los valores a escribir en los registros:  
LED2_OFF_H = 0x3;  
LED2_OFF = 0x32;

## Cuarto PWM  
60% de retardo de incio, ciclo util de 20% y tiempo de apagado de 80%  
LED3ON=round(4096*.60)-1  

In [23]:
LED3ON=round(4096*.60)-1  
print(LED3ON)
np.base_repr(LED3ON,16)

2457


'999'

LED3_ON_H = 0x9;  
LED3_ON_L = 0x99;  

Valores para LED3_OFF:
EL tiempo encendido sera entonces de 80% -> 3276.8 counts con un apagado de 20% por lo que LED2_OFF = ((2457+3276)=val-4096)

In [24]:
value6 = resolution * .80
LED3OFF=((round(LED3ON)+round(value6))-4096)
print(LED3OFF)
np.base_repr(LED3OFF,16)

1638


'666'

LED3_OFF_H = 0x6;  
LED3_OFF = 0x66;